[Done] Monster speed overhaul

kevingranade opened the pull request:
Incidental bugfix for the unit test, plus rebalance for stumbling monster speeds.
After I made monsters stumble much more pervasively, they're effectively slowed down a great deal, so I wrote this unit test to check what their effective speed is now.  as it turns out, it's VERY slow, zombies have an effective speed of 58, even zombie dogs have an effective speed much slower than the player at a walk, and while super-fast things like Jabberwocks and hulks are faster than the player at a walk, they can easily be outdistanced if the player simply runs for a minute or so.

The test is very much a simulation, it creates a featureless plain, plunks down the monster on it, and gives the monster a goal 100 squares away, then runs through minimal monster turns until the monster arrives, finally returning the number of turns taken.

The values at the bottom are the keystone speed values I want to test for (I guess I should add the low end, crawlers, but they're so slow it hardly matters), zombies should be significantly slower than the player, but capable of catching up if the player is impaired, zombears I envision as having a slight edge over the player at a walk, but are easily outdistanced by running, zombie dogs are the hunters of the horde, it should be difficult, but possible for the player to outrun them, and finally Jabberwocks (and Hulks) should be extremely difficult to outrun.  This is actually backing off a bit from the status quo, which was that there were a number of monsters faster than the player, but with the way shambling monsters avoid obstacles (accidentally, but it's still effective), it's no longer a viable tactic to lead pursuers over shrubs to lose them.

I think these are reasonable goals, keeping in mind an "average" player has speed 100, and can briefly double their speed by running, but obviously I'm posting to discuss, so let's do that.
Also keep in mind that if someone wants the "romero slow zombie" type experience, we're better served by telling them to use the slow zombies mod than trying to make the main game cover both bases.  Likewise this will heavily affect the fast zombies mod, but adjusting their balance is a separate issue I think.

A side note is that buffing zombie speeds to reach these movement speed goals will have them attacking VERY quickly in some cases (2-3x per player turn with the fastest monsters), that might be something that needs to be reigned in, though at the low end there are certainly a number of monsters that could use the buff.

kevingranade commented:
Another feature of this PR is that it has addressed several issues that were blocking handling monsters in a unit test, we can now write tests where we spawn monsters and interact with them, for example to test lethality of various weapons.

BevapDin commented:
I get this result. It seems to assume CIRCLEDIST is set to `true` (its default value is `false`, and that's the value it still has in my configuration)
```
monster_test.cpp:119: FAILED:
  CHECK( diag_move >= 141 - 3 )
with expansion:
  99 >= 138

monster_test.cpp:83: FAILED:
  CHECK( horizontal_move.avg() <= expected_move_points + 4 )
with expansion:
  188 <= 129.0f
with message:
  horizontal_move.avg() := 188

monster_test.cpp:87: FAILED:
  CHECK( vertical_move.avg() <= expected_move_points + 4 )
with expansion:
  185 <= 129.0f

monster_test.cpp:91: FAILED:
  CHECK( diagonal_move.avg() <= (expected_move_points * 1.63) + 4 )
with expansion:
  282 <= 207.75

monster_test.cpp:83: FAILED:
  CHECK( horizontal_move.avg() <= expected_move_points + 4 )
with expansion:
  108 <= 94.90909f
with message:
  horizontal_move.avg() := 108

monster_test.cpp:87: FAILED:
  CHECK( vertical_move.avg() <= expected_move_points + 4 )
with expansion:
  109 <= 94.90909f

monster_test.cpp:91: FAILED:
  CHECK( diagonal_move.avg() <= (expected_move_points * 1.63) + 4 )
with expansion:
  166 <= 152.1818136597

monster_test.cpp:83: FAILED:
  CHECK( horizontal_move.avg() <= expected_move_points + 4 )
with expansion:
  126 <= 70.66666f
with message:
  horizontal_move.avg() := 126

monster_test.cpp:87: FAILED:
  CHECK( vertical_move.avg() <= expected_move_points + 4 )
with expansion:
  125 <= 70.66666f

monster_test.cpp:91: FAILED:
  CHECK( diagonal_move.avg() <= (expected_move_points * 1.63) + 4 )
with expansion:
  188 <= 112.6666625214

monster_test.cpp:83: FAILED:
  CHECK( horizontal_move.avg() <= expected_move_points + 4 )
with expansion:
  95 <= 59.55556f
with message:
  horizontal_move.avg() := 95

monster_test.cpp:87: FAILED:
  CHECK( vertical_move.avg() <= expected_move_points + 4 )
with expansion:
  95 <= 59.55556f

monster_test.cpp:91: FAILED:
  CHECK( diagonal_move.avg() <= (expected_move_points * 1.63) + 4 )
with expansion:
  143 <= 94.5555583191
```

And with cicrcular distance set to true, I get those:
```
monster_test.cpp:83: FAILED:
  CHECK( horizontal_move.avg() <= expected_move_points + 4 )
with expansion:
  245 <= 129.0f
with message:
  horizontal_move.avg() := 245

monster_test.cpp:87: FAILED:
  CHECK( vertical_move.avg() <= expected_move_points + 4 )
with expansion:
  247 <= 129.0f

monster_test.cpp:91: FAILED:
  CHECK( diagonal_move.avg() <= (expected_move_points * 1.63) + 4 )
with expansion:
  407 <= 207.75

monster_test.cpp:83: FAILED:
  CHECK( horizontal_move.avg() <= expected_move_points + 4 )
with expansion:
  144 <= 94.90909f
with message:
  horizontal_move.avg() := 144

monster_test.cpp:87: FAILED:
  CHECK( vertical_move.avg() <= expected_move_points + 4 )
with expansion:
  143 <= 94.90909f

monster_test.cpp:91: FAILED:
  CHECK( diagonal_move.avg() <= (expected_move_points * 1.63) + 4 )
with expansion:
  227 <= 152.1818136597

monster_test.cpp:83: FAILED:
  CHECK( horizontal_move.avg() <= expected_move_points + 4 )
with expansion:
  161 <= 70.66666f
with message:
  horizontal_move.avg() := 161

monster_test.cpp:87: FAILED:
  CHECK( vertical_move.avg() <= expected_move_points + 4 )
with expansion:
  163 <= 70.66666f

monster_test.cpp:91: FAILED:
  CHECK( diagonal_move.avg() <= (expected_move_points * 1.63) + 4 )
with expansion:
  262 <= 112.6666625214

monster_test.cpp:83: FAILED:
  CHECK( horizontal_move.avg() <= expected_move_points + 4 )
with expansion:
  123 <= 59.55556f
with message:
  horizontal_move.avg() := 123

monster_test.cpp:87: FAILED:
  CHECK( vertical_move.avg() <= expected_move_points + 4 )
with expansion:
  124 <= 59.55556f

monster_test.cpp:91: FAILED:
  CHECK( diagonal_move.avg() <= (expected_move_points * 1.63) + 4 )
with expansion:
  203 <= 94.5555583191
```

The numbers are fairly consistent over several runs. Btw. it might be useful to explicitly initialize the RNG to some default value in the test program.

kevingranade commented:
The test is adjusted, but the actual speed values of the monsters are not,
so they are expected to fail.  I should have called that out explicitly.
Thanks for pointing out the circular distance thing though, that needs to
be set explicitly by the test, I was assuming it was using the option
defaults.

i2amroy commented:
Okay, couple of questions here, just to get some things sorted out about the "rebalancing":
1) Are we looking at circular distances? I know that personally I don't play with them, and from what I've seen the significant majority of the player base doesn't either (most likely because they are off by default).
2) If we are is it worth taking into effect the difference for balancing requirements there by having balancing notions that kick in to help account for the lesser one? As I see it the values in the JSON should be balanced towards the default (most common) setting, and option-based changes should be either just assumed to be part of the option change, or should be addressed separately in the code.

As a potential solution, rather than tweaking speed vastly could we look at potentially reducing the "move cost" of tiles by either a flat value or percentage when they stumble instead? (i.e. Normally moving over flat ground costs 100 moves per tile, but instead have stumbling moves only cost 80). That seems like it would allow for keeping speed the same (and thus attack speed) while allowing for an essentially increased movement speed.

kevingranade commented:
1) The test assumes circular distances, and the impact of stumbling on
non-circular distances should be somewhat minimal, so there definitely
needs to be some kind of tweak there, because rl dist zombies will be
effectively 50%+ faster than circular dist zombies.
1a) I keep meaning to make circular the default, but that's not a question
for this PR.
2) Yes, it needs to be balanced for both, ideally by making the speed the
same for both.
2a) I don't like the idea of zombies moving faster just when they stumble,
but another way to get a similar effect is to make just move speed
effectively faster by giving monsers a discounted number of moves per
square if they are stumbling type monsters and trig distance is on.

kevingranade commented:
More consisely, I want the effective speed (with stumbling taken into
account) to be the same with both distance options.  Stated that way,
adjusting move pts/square instead of adjusting monster speed is the better
option.

i2amroy commented:
>1a) I keep meaning to make circular the default, but that's not a question
for this PR.

Probably good that we aren't discussing that here, since I have the exact opposite opinion. :P

>Stated that way, adjusting move pts/square instead of adjusting monster speed is the better
option.

+1 there, seems like the best option to me.

Rivet-the-Zombie commented:
>circular distance default

I'm in favor of this.

>adjusting move point cost instead of monster speed

This too.

Coolthulhu commented:
> circular the default

-1

> adjusting move pts/square instead of adjusting monster speed is the better option

+1, but zombies could also get a speed buff, now that player can easily outrun hulks and hunters. Maybe also decoupling movement speed from global speed, so that all zeds could be just speed 90-110 (~1 attack per turn), but move at different speeds (no reason for hulk to attack 1.4 times per turn).

illi-kun commented:
> circular the default

+1!

i2amroy commented:
I'm just going to make a separate issue for the circular/non-circular thing to get it out of here. :P

>Maybe also decoupling movement speed from global speed, so that all zeds could be just speed 90-110 (~1 attack per turn), but move at different speeds (no reason for hulk to attack 1.4 times per turn).

Splitting attack speed and movement speed is always a nice thing in my book, but I also really like the fact that code wise every action uses the same units. If we were going to go that route what we should probably do is maintain the old speed token but add two new modifiers instead.

For example you could then drop Hulk "speed" (the total number of move points) down to 100 like most other things, but then they could have "move_speed": 80, meaning that it only costs them 80 move points per 100 the tile normally costs, and "attack_speed": 110, making it cost them 110 points to throw a punch.

In that way altering "speed" would be reserved for improvements that affect everything (crafting, item use, etc.) while "move_speed" and "attack_speed" would specifically only apply to the movement and attack costs alone.

kevingranade commented:
Another option is having 'speed' act as it does now, but add 'attack_cost',
which determines # of moves per attack.  As a bonus it could default to
being whatever it is now, so it can be implemented without adding it to
everything.

kevingranade commented:
Instead of adjusting monster speed, I adjusted move cost, bringing it in line with the monster's stated speed.
It turns out there's an odd effect in roguelike distance, where as the angle the monster is moving at approaches 45 degrees, the monsters effective speed would drop significantly, hence passing the angle from monster::move() to monster::move_to() to adjust for that, now however effective speed is the same in all directions, to +/- 4%.
There was also a bug where when using roguelike distance, stumbling monsters would line up like they did before, I was able to work around it by using floating point circular distance measurements to weight stumbling direction.
Along the way I added the attack_cost that we discussed, though I didn't end up needing it.  We can remove it, but I think we wanted to adjust attack speed independently of move speed anyway, I'm just blanking on a monster to apply it to at the moment.  After I made that last comment I realized i2amroy and I were saying the same thing, durr.

Coolthulhu commented:
Mingw doesn't like the class name `stat`. This only breaks the tests, but it does break the tests.

Coolthulhu commented:
Isn't very consistent when it comes to circular vs square.

Typical situation in game:
* Flat, featureless map
* Zombie dog (speed 130)
* Chasing a quick, unencumbered survior (110 move cost)

Results:
* In square mode, the results are as expected: dog gets a hit in once in a while, regardless of player's movement direction
* In circular mode, when player is moving diagonally, dog is getting hits once in a while (approximately one every 4 steps)
* In circular mode, when player is moving orthogonally, the player is obviously gaining distance

Maybe there should be a separate test simulating a moving target that stays relatively close to the monster?

kevingranade commented:
I have a test that reproduces the issue, but taking another look, zombie dogs have a speed of 105, not 130, so it's right at the cusp of player speed, which is the worst case for balancing :/
Still needs to be fixed of course, the main thing that needs to happen is there should be no benefit to running from zombies in a particular direction.

Coolthulhu commented:
I see that last commit here is pretty old.
I didn't merge because it has some flaws, but I'd say it is better than what we have now. How about I'd merge it now and fixes would be separate PRs?

kevingranade commented:
What I have locally corrects it a little bit, so I'll push that if I get to
it before you merge.

Coolthulhu commented:
Note: breaks tests on mingw. Just the tests.
